[MessageBoard] Add TS Demo for Message board

areshand opened the pull request:
<!--
Thank you for sending a PR. We appreciate you spending time to help improve the Aptos Core project.

The project is undergoing daily changes. Pull Requests will be reviewed and responded to as time permits.
-->

## Motivation

1. Demo the ACL messageboard with basic JS (**messageboard_demo/client/index.ts**)
2. Add Cli option to move_example for compiling move with named addresses
3. add ACL.move and Offer.move to messageboard since move nursery is not ready to be included in move stdlib yet. 

followup: (1) add demo for cap based board

### Have you read the [Contributing Guidelines on pull requests](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan
cargo test -p framework

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/aptos-labs/aptos-core/tree/main/developer-docs-site, and link to your PR here.)

CapCap commented on developer-docs-site/static/move_demo/client/index.js:
```diff
@@ -0,0 +1,115 @@
+const aptos = require("aptos");
```
could you use TS for this please, like some of the other demos?

areshand replied:
addressed the comments

davidiw commented on aptos-move/move-examples/messageboard/sources/Offer.move:
```diff
@@ -0,0 +1,138 @@
+/// Provides a way to transfer structs from one account to another in two transactions.
```
I feel like this should just be part of the standard library if you want to incorporate it.

areshand replied:
added to stdlib

davidiw commented on aptos-move/move-examples/messageboard/sources/ACL.move:
```diff
@@ -0,0 +1,46 @@
+/// Access control list (ACL) module. An ACL is a list of account addresses who
```
same, as offer, if we're going to recommend using this, we should just make it generally available. That would also reduce burden on the chain, no?

areshand replied:
added to stdlib

davidiw commented on aptos-move/move-examples/messageboard/sources/Offer.move:
```diff
@@ -0,0 +1,138 @@
+      recipient == global<Offer<Offered>>(offer_addr).for || recipient == offer_addr
+    }
+  }
+
+
+
```
wow, so many trees lol

areshand replied:
removed

davidiw commented on developer-docs-site/static/messageboard_demo/client/index.ts:
```diff
@@ -0,0 +1,172 @@
+        return payload;
+    }
+
+    async createMessageboard(account: AptosAccount, modulePath: string) {
```
this should be two different steps....

Ideally we should have "installMessageboardModule" and "createMessageboard" separately.

installation happens only once, but there's no reason why we would ever need more than one instance of the module installed. Whereas Alice and Bob can use the same module but call "createMessageboard" using the same backend. Much like we did in the first move module tutorial.

areshand replied:
split to two

davidiw reviewed (changes requested):
Really great start to an interactive app. It might be great to specify a little bit more here and actually build out a slightly more full fledged app that:
* keeps track of events that have been seen, so they aren't pulled again
* supports remove participant
* keeps a list of historical participants -- maybe even all possible participants in the case of the capability and acl tables -- to support removal.

With that being said, we've done a lot of work on this already, so it may be fine to just clean it up a bit with a solid test and pause effort here for a little bit.

This demo does something dangerous, which it teaches the user to install a module each time they wish to instantiate a resource. The problem with installing the same module over and over again is that someone would inevitably have to build an app that can guess that the module is effectively the same just installed in different places.

Also the code moves around the hello blockchain code which is reference in the first move module tutorial but it doesn't update the tutorial.

davidiw commented on developer-docs-site/static/messageboard_demo/client/index.ts:
```diff
@@ -0,0 +1,172 @@
+    // compile the modules with the admin's address
+    chdir('../../../../aptos-move/move-examples');
+    exec(
+        `cargo run -- messageboard/sources/ -a MessageBoard -h ${admin.address().toString()}`,
```
this is really cool, we should update the documentation for first move module asap.

areshand commented:
> 



> Really great start to an interactive app. It might be great to specify a little bit more here and actually build out a slightly more full fledged app that:
> 
> * keeps track of events that have been seen, so they aren't pulled again
currently, the SDK doesn't support  passing page or limit parameter. 
will support this feature once SDK supports passing page or limit parameters. 
> * supports remove participant
added

> * keeps a list of historical participants -- maybe even all possible participants in the case of the capability and acl tables -- to support removal.
added some basic tracking of participants. will refine when adding cap based board

> 
> With that being said, we've done a lot of work on this already, so it may be fine to just clean it up a bit with a solid test and pause effort here for a little bit.
> 
> This demo does something dangerous, which it teaches the user to install a module each time they wish to instantiate a resource. The problem with installing the same module over and over again is that someone would inevitably have to build an app that can guess that the module is effectively the same just installed in different places.

Yea, great callout. separate it to two functions to not misleading the reader. 
> 
> Also the code moves around the hello blockchain code which is reference in the first move module tutorial but it doesn't update the tutorial.

hmm, seems the move file path is not used in code. seems the website is displaying correctly https://deploy-preview-546--aptos-developer-docs.netlify.app/tutorials/your-first-move-module. am i missing something here?

davidiw commented on aptos-move/move-examples/src/main.rs:
```diff
@@ -1,28 +1,38 @@
         generate_abis: false,
         generate_docs: true,
         install_dir: args.output_path,
+        additional_named_addresses: named_address,
```
@gregnazario fyi

gregnazario replied:
Yeah, this is already built into the CLI

davidiw replied:
oh, I see that @areshand did: `MessageBoard = "_"`

davidiw reviewed (commented):
this is really cool

1) update first move module tutorial due to new paths
2) use the cool new parameter to make it easier to use named addresses -- no more of that nonsensical update the Move.toml file
3) consider consulting with Greg to port the compilation to the Aptos CLI
4) Do we have any readme or guide for this?

code looks good to me -- but it would break the existing documentation for first move module.

areshand commented:
@davidiw  updated the tutorial so it is not broken. for the readme, will add after this lands so that I can link to the code in the readme more easily

davidiw commented on aptos-move/move-examples/hello_blockchain/Move.toml:
```diff
@@ -0,0 +1,9 @@
+version = "0.0.0"
+
+[addresses]
+HelloBlockchain = "66301a58029b93bb46e93d344a79d5673a62c81f94a829b93d0e548a19269fdf"
```
can you set this to "_" like you did for message board? and maybe update the instructions so folks can compile this more easily

areshand replied:
modified

davidiw reviewed (approved):
woot woot

areshand commented:
/land

areshand commented:
/land
